[MRG] Raise descriptive ValueError if number of samples equals number of classes in Linear Discriminant Analysis

eddbrown opened the pull request:
Fixes #12374

Before: Failed with division by zero error if no_classes == no_sample in Linear Discriminant Analysis.

After: This PR raises a more descriptive ValueError, explaining the problem.

<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->



<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

eddbrown committed:
Raise descriptive value error if number of classes equals the number of samples

jnothman commented:
Please give your PR a more descriptive title

jnothman commented:
The PR *description* should contain "Fixes #xxxx"

adrinjalali commented:
Could you please also add a test? You also have a PEP8 issue, hence travis failing. `flake8 /path/to/file.py` would tell you where to fix as well.

adrinjalali commented on sklearn/tests/test_discriminant_analysis.py:
```diff
@@ -350,3 +350,15 @@ def test_covariance():
+    X = np.zeros((no_classes, no_classes))
+    y = np.zeros((no_classes, 1))
+    clf = LinearDiscriminantAnalysis(solver="svd")
+    assert_raises(ValueError, clf.fit, X, y)
```
```suggestion
    with pytest.raises(ValueError, match="The number of samples must be more"):
        clf.fit(X, y)
```

eddbrown replied:
ah nice, tests the message too to make sure that its actually THAT ValueError being raised

adrinjalali replied:
yes, and since the adoption of `pytest`, `pytest....` are generally preferred to custom `assert_...` ones.

eddbrown replied:
thanks for your advice!

jnothman commented:
Tests are failing

jnothman commented:
Please append commits rather than amending and force-pushing. It makes it hard for us to review when we aren't notified that a commit has been added and we can't see the diff

jnothman reviewed (commented):
Since classes_ is set as unique values in y, it seems a little strange to have the message state the number of samples and classes when they must be equal

adrinjalali commented on sklearn/discriminant_analysis.py:
```diff
@@ -358,6 +358,13 @@ def _solve_svd(self, X, y):
+                             "Currently, you have {} samples and {} "
+                             "classes.".format(n_samples, n_classes))
+
+
```
please have only one empty new line.

jnothman commented:
This case should also be handled for the other solvers. Currently eigen raises an error, while lsqr gives a warning.

Commit:
Cover all linear discriminant analysis methods

The lsqr and eigen solvers also will fail if the number of samples equals the number of classes, so this commit moves the error raise to cover all cases.

Commit:
Add back missing variables

eddbrown commented:
@jnothman @adrinjalali thanks for your patience. How is it looking?

jnothman commented on sklearn/tests/test_discriminant_analysis.py:
```diff
@@ -350,3 +352,15 @@ def test_covariance():
     assert_almost_equal(c_s, c_s.T)
+
+
+def test_raises_value_error_on_same_number_of_classes_and_samples():
```
Use pytest.mark.parameterize to test this for each solver I suppose

jnothman reviewed (commented):
Otherwise LGTM. I don't know if it's helpful to users to put this in our change log...?

Commit:
Iterate through solvers with pytest parameterise

Commit:
Fix test

adrinjalali reviewed (approved):
Thanks @eddbrown , LGTM.

eddbrown commented:
@adrinjalali what is the next step with getting this merged then? I think I've misunderstood the process a bit...

adrinjalali commented:
@eddbrown an unwritten part of the process is the casual (read usual, normal) patience required for a core developer to do the next action ;) let it be a review, response to a question, merge, etc.

Your PR now is in a state to be merged, and I guess it will be soon (I'm not somebody with permissions to do a merge). Now you can just sit back and relax until it happens, or go on and find another issue to investigate ;)

jnothman commented:
>  an unwritten part of the process is the casual (read usual, normal) patience 

I'm pretty sure I mentioned patience somewhere in the contributor docs in the last couple of years... ?

Thanks @eddbrown
